### PR TITLE
Clear completion pattern matcher caches on solution close

### PR DESCRIPTION
**Pull request by rchande**
Adds a per-workspace service that can provide CompletionHelpers and that clears them when that workspace's solution is closed.

The CompletionHelper.*Instance fields regularly get flagged by RPS as regressions because they're still around after "Close Solution". With this change, the caches should stop showing up.

Tag @dotnet/roslyn-ide

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Lulz

File: src/Features/Core/Portable/Completion/CompletionService.cs [hunk 1]
```diff
@@ -135,7 +135,7 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
             ImmutableArray<CompletionItem> items,
             string filterText)
         {
-            var helper = CompletionHelper.GetHelper(document);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionService.cs]: couldn't you just update this method to do exactly what you've expanded out all the code to do?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs [hunk 2]
```diff
@@ -267,7 +267,8 @@ private Document GetDocument()
         private CompletionHelper GetCompletionHelper()
         {
             var document = GetDocument();
-            return CompletionHelper.GetHelper(document);
+            var service = document.Project.Solution.Workspace.Services.GetService<ICompletionHelperService>();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: you've repeated this like 10 times.  Why not just have CmpletionHelper.GetHelper(Document) do exactly this?

File: src/Features/Core/Portable/Completion/CompletionHelperServiceFactory.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            return new Service(workspaceServices.Workspace);
+        }
+
+        private class Service : ICompletionHelperService, IWorkspaceService
+        {
+            private object _gate = new object();
+
+            private CompletionHelper caseSensitiveInstance;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionHelperServiceFactory.cs]: underscores.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs (hunk 2 above)
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs]: Good point, will do.

File: src/Features/Core/Portable/Completion/CompletionHelperServiceFactory.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    internal class CompletionHelperServiceFactory : IWorkspaceServiceFactory
+    {
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            return new Service(workspaceServices.Workspace);
+        }
+
+        private class Service : ICompletionHelperService, IWorkspaceService
+        {
+            private object _gate = new object();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionHelperServiceFactory.cs]: make readonly.

**Comment by rchande** [comment]
Tag @MattGertz for approval

**Customer scenario**

The user opens VS and works in a C#/VB solution. Without exiting VS, they close their solution and start working on a C++ or JS project. Roslyn's Completion PatternMatch caches are still held in memory because they are rooted in static fields. This could potentially degrade the user's experience.

**Bugs this fixes:**

https://devdiv.visualstudio.com/web/wi.aspx?pcguid=011b8bdf-6d56-4f87-be0d-0092136884d9&id=450387&projectId=0bdbc590-a062-4c3f-b0f6-9383f67865ee&src=WorkItemMention&src-action=artifact_link&fullScreen=true

**Workarounds, if any**

None

**Risk**

Low--the PatternMatch caches are now associated with a Workspace and are cleared out when the solution is closed.

**Performance impact**

There should be no impact to the user while working in managed projects, they may see a better experience in non-managed projects. 

**Is this a regression from a previous update?**

RPS claims this is a regression

**Root cause analysis:**

We added this cache to improve the typing experience but did not consider scenarios where users opened different project types without closing Visual Studio first.

**How was the bug found?**

RPS

**Comment by rchande** [comment]
Looks like VSI tests passed but xunit hit appdomainunloaded at the end of the test run.

**Pull request merged**